tracks/python/exercises/palindrome-products/mentoring.md

```diff
@@ -0,0 +1,188 @@
+# Mentoring
+
+## Problem and challenges
+
+The problem asks for two functions, `smallest` and `largest`, that will respectively give the smallest and the largest palindrome number that is a product of at least one pair of factors in a given range.
+
+The main challenge of this problem is performance.
+A naive **O(N²logN)** solution can lead to timeouts when running the tests online.
+
+## Reasonable solution
+
+```python
+from typing import Callable, Optional, Sequence
+
+Factors = list[list[int]]
+Result = tuple[Optional[int], Factors]
+Comparison = Callable[[int, int], bool]
+
+
+def is_palindrome(num: int) -> bool:
+    textual = str(num)
+    return textual == textual[::-1]
+
+
+def search(factor_range: Sequence[int],
+           better_found: Comparison) -> Result:
+    factor_range = list(factor_range)
+    found = None
+    factors: Factors = []
+    for i in factor_range:
+        for j in factor_range:
+            num = i*j
+            # note: further values of j will yield even worse num
+            if found and not better_found(num, found):
+                break
+            if is_palindrome(num):
+                if num == found:
+                    factors.append([i, j])
+                else:
+                    factors = [[i, j]]
+                found = num
+    return (found, factors)
+
+
+def largest(min_factor: int, max_factor: int) -> Result:
+    if min_factor > max_factor:
+        raise ValueError('min must be <= max')
+    # decrementing for time optimization (see note above)
+    return search(range(max_factor, min_factor - 1, -1),
+                  lambda num, found: num >= found)
```

Comment:
This already is defined for you: https://docs.python.org/3/library/operator.html#operator.ge